src/main/scala/com/fulcrumgenomics/bam/DownsampleAndNormalizeBam.scala

```diff
@@ -0,0 +1,206 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.bam.DownsampleAndNormalizeBam.CoverageTracker
+import com.fulcrumgenomics.bam.api._
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.commons.util.LazyLogging
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.util.{Io, ProgressLogger, Sorter}
+import htsjdk.samtools.util.IntervalList.IntervalMergerIterator
+import htsjdk.samtools.util.{CoordMath, Interval, IntervalList, Murmur3, OverlapDetector}
+
+/**
+  * Companion object with helper classes.
+  */
+private object DownsampleAndNormalizeBam extends LazyLogging {
+  /**
+    * Tracks coverage over a set of regions.
+    */
+  class CoverageTracker(regions: Seq[Interval],
+                        val targetCoverage: Int,
+                        val minMapQ: Int
+                       ) {
+
+    // Overlap detector for region => coverage
+    private val detector = new OverlapDetector[RegionCoverage](0, 0)
+    regions.foreach { region =>
+      detector.addLhs(new RegionCoverage(region.getContig, region.getStart, region.getEnd), region)
+    }
+
+    /**
+      * Checks to see if a template has reads that contribute coverage to any bases that are currently
+      * below the coverage threshold.  If so, coverage from all acceptable reads is accumulated and the
+      * method returns true.  Otherwise no coverage is accumulated and the method returns false.
+      *
+      * @param t a Template to accumulate coverage from if it covers any bases that are under the coverage threshold
+      * @return true if the template added coverage, false otherwise
+      */
+    def add(t: Template): Boolean = {
+      val recs = t.allReads
+        .filterNot(_.secondary)
+        .filterNot(_.unmapped)
+        .filterNot(_.duplicate)
+        .filter(_.mapq >= minMapQ)
+        .toSeq
+
+      val addsCoverage = recs.exists { rec =>
+        detector.getOverlaps(rec.asSam).iterator().exists { cov =>
+          Range.inclusive(math.max(rec.start, cov.start), math.min(rec.end, cov.end)).exists { pos =>
+            cov(pos) <= targetCoverage
```

Comment:
I think we only want to add the record if the "if a template has reads that contribute coverage to any bases that are currently below the coverage threshold":
```suggestion
            cov(pos) < targetCoverage
```